src/dxvk/dxvk_context.cpp

```diff
@@ -2609,6 +2610,129 @@ namespace dxvk {
     m_cmd->trackResource<DxvkAccess::None>(event);
   }
   
+
+  void DxvkContext::launchCuKernel(VkCuLaunchInfoNVX& launchInfo, const std::vector<Rc<DxvkBuffer>>& readBuffers, const std::vector<Rc<DxvkBuffer>>& writeBuffers, const std::vector<Rc<DxvkImage>>& readImages, const std::vector<Rc<DxvkImage>>& writeImages)
+  {
+    // the resources in the std::vectors above are called-out
+    // explicitly in the API for barrier and tracking purposes
+    // since they're being used bindlessly.
+    //
+    // resources may appear in the input read *and* write sets
+    // simultaneously, so start by making a joint read-write set
+    // so each resource is only transitioned once.  note: in
+    // practice there are typically just a handful of resources.
+    std::unordered_set<DxvkBuffer*> allBuffers(readBuffers.size() + writeBuffers.size());
+    std::unordered_set<DxvkImage*> allImages(readImages.size() + writeImages.size());
+    for (auto& r : readBuffers) {
+      allBuffers.insert(r.ptr());
+    }
+    for (auto& r : writeBuffers) {
+      allBuffers.insert(r.ptr());
+    }
+    for (auto& r : readImages) {
+      allImages.insert(r.ptr());
+    }
+    for (auto& r : writeImages) {
+      allImages.insert(r.ptr());
+    }
```

Comment:
I'm happy to do that but I'm not sure how that'd affect the resulting barriers - do you mean that if an image is in both the readable and writable lists it's better to just treat it as DxvkAccess::Write only?  Or am I misunderstanding?

Reply:
There's two concerns here:
a) we're doing an additional heap allocation just to combine two lists into one, when none of the code really relies on the two lists being separate in the first place. I'd strongly prefer the `(image,access)` tuples since that removes the need for pretty much all of this.
b) wrt barriers, we're currently treating *all* resources as read-write, which means that if a resource from the read list has previously been used in a read-only way as well (some sort of compute shader, copy, you name it), we may insert an unnecessary barrier anyway.

Same goes for buffers, fwiw.

Reply:
To explain where the lists are coming from - they're coming from the launch entrypoint in NvAPI, which get two arrays (read, write) of opaque object handles... which in reality for DXVK are two lists of ID3D11Resource*s but not differentiated by image-vs-buffer.  Someone somewhere along the chain should ref them before submission to the CS thread, detect entries that are in both lists (to avoid double-transition) and differentiate image vs buffer.

I could do more of this before we get to launchCuKernelNVX(), likely remove one of the steps, though it's attractive (to me anyway) to take the bulk off of the application thread.

What d'ya think?  I'll look at this more deeply tomorrow, see what seems relatively elegant.

Reply:
> I could do more of this before we get to launchCuKernelNVX(), likely remove one of the steps, though it's attractive (to me anyway) to keep the bulk off of the application thread.

That's what I mean, we're doing a fair bit of work there anyway so might as well process the two lists in a way that is more easy for the backend to work with. If duplicates can occur, even a naive O(n²) algorithm would be fine as long as the number of resources of each type that we're dealing with is no more than ~a dozen or so, this is how our barrier code works as well.

The primary bottleneck should be our worker thread anyway rather than the app thread, as long as the app doesn't spend more time in its own code than it does submitting D3D commands.

Reply:
Thinking about building the ID3D11Resource*s straight into these vectors (or sets, but probably vectors) in D3D11DeviceContextExt::LaunchCubinShaderNVX() and then using them as-is in DxvkContext::launchCuKernelNVX()
```
    std::vector<std::tuple<Rc<DxvkImage>, bool /*read*/, bool /*write*/>> images;
    std::vector<std::tuple<Rc<DxvkBuffer>, bool /*read*/, bool /*write*/>> buffers;
```

Reply:
Oops, didn't mean to resolve.
I've been pursuing this approach though and it seems to be working out neatly.

Reply:
I've pushed the rejig of this.  There are only two lists and they're only built once now, and then deduped in-place later on the CS thread.  The barriers are now finer-grained.

I haven't verified yet that this is doing what I expect (I will - next), just that nothing seems broken!